File: code.js
class Rectangle {
    constructor(x, y, width, height) {
        this.x = x;
        this.y = y;
        this.width = width;
        this.height = height;

        this.top = y;
        this.bottom = y + height;
        this.left = x;
        this.right = x + width;
    }

    contains(x, y) {
        return this.left < x && x < this.right && this.top < y && y < this.bottom;
    }

    moveTo(x, y) {
        this.x = x;
        this.y = y;

        this.top = y;
        this.bottom = y + this.height;
        this.left = x;
        this.right = x + this.width;
    }

    resizeTo(handle, x, y) {
        switch (handle) {
            case 'left':
                this.x = x;
                this.left = x;
                this.width = this.right - this.left;
                if (this.width < 5) {
                    this.width = 5;
                    this.left = this.right - 5;
                    this.x = this.left;
                }
                break;
            case 'right':
                this.right = x;
                this.width = this.right - this.left;
                if (this.width < 5) {
                    this.width = 5;
                    this.right = this.left + 5;
                }
                break;
            case 'top':
                this.y = y;
                this.top = y;
                this.height = this.bottom - this.top;
                if (this.height < 5) {
                    this.height = 5;
                    this.top = this.bottom - 5;
                    this.y = this.top;
                }
                break;
            case 'bottom':
                this.bottom = y;
                this.height = this.bottom - this.top;
                if (this.height < 5) {
                    this.height = 5;
                    this.bottom = this.top + 5;
                }
                break;
            case 'topleft':
                this.y = y;
                this.top = y;
                this.height = this.bottom - this.top;
                if (this.height < 5) {
                    this.height = 5;
                    this.top = this.bottom - 5;
                    this.y = this.top;
                }
                this.x = x;
                this.left = x;
                this.width = this.right - this.left;
                if (this.width < 5) {
                    this.width = 5;
                    this.left = this.right - 5;
                    this.x = this.left;
                }
                break;
            case 'bottomleft':
                this.bottom = y;
                this.height = this.bottom - this.top;
                if (this.height < 5) {
                    this.height = 5;
                    this.bottom = this.top + 5;
                }
                this.x = x;
                this.left = x;
                this.width = this.right - this.left;
                if (this.width < 5) {
                    this.width = 5;
                    this.left = this.right - 5;
                    this.x = this.left;
                }
                break;
            case 'topright':
                this.y = y;
                this.top = y;
                this.height = this.bottom - this.top;
                if (this.height < 5) {
                    this.height = 5;
                    this.top = this.bottom - 5;
                    this.y = this.top;
                }
                this.right = x;
                this.width = this.right - this.left;
                if (this.width < 5) {
                    this.width = 5;
                    this.right = this.left + 5;
                }
                break;
            case 'bottomright':
                this.bottom = y;
                this.height = this.bottom - this.top;
                if (this.height < 5) {
                    this.height = 5;
                    this.bottom = this.top + 5;
                }
                this.right = x;
                this.width = this.right - this.left;
                if (this.width < 5) {
                    this.width = 5;
                    this.right = this.left + 5;
                }
                break;
        }

    }
}

function drawRectangle(ctx, rectangle) {
    ctx.strokeRect(rectangle.x, rectangle.y, rectangle.width, rectangle.height);
}

function addContextMenu(canvas, callbacks) {
    const contextMenu = document.getElementById("context-menu");
    const closeContextMenu = () => {
        contextMenu.style.display = "none";
        while (contextMenu.firstChild) {
            contextMenu.removeChild(contextMenu.lastChild);
        }
    };
    canvas.addEventListener("contextmenu", (e) => {
        console.log(e);
        contextMenu.style.top = e.clientY + "px";
        contextMenu.style.left = e.clientX + "px";
        contextMenu.style.display = "block";

        const createNewBlockButton = createButtonCreateNewBlock(() => {
            callbacks.onCreateNewBlock(e.clientX, e.clientY);
            closeContextMenu();
        });

        contextMenu.appendChild(createNewBlockButton);

        e.preventDefault();
    });
    canvas.addEventListener("click", closeContextMenu);
}

function createButtonCreateNewBlock(onCreateNewBlock) {
    const button = document.createElement('button');
    button.textContent = 'Create new block';
    button.addEventListener('click', (e) => {
        onCreateNewBlock();
        e.preventDefault();
    });
    return button;
}


function findBlock(blocks, x, y) {
    for (const block of blocks) {
        if (block.contains(x, y)) {
            return block;
        }
    }
    return null;
}

class DraggingRectangle {
    constructor(rectangle, offsetX, offsetY) {
        this.rectangle = rectangle;
        this.offsetX = offsetX;
        this.offsetY = offsetY;
    }

    moveTo(x, y) {
        this.rectangle.moveTo(x - this.offsetX, y - this.offsetY);
    }
}

function addDragFeature(canvas, blocks, redraw) {
    var draggableBlock = null;

    canvas.addEventListener('mousedown', (e) => {
        const mouseX = e.clientX - canvas.getBoundingClientRect().left;
        const mouseY = e.clientY - canvas.getBoundingClientRect().top;

        const block = findBlock(blocks, mouseX, mouseY)

        if (block) {
            draggableBlock = new DraggingRectangle(
                block,
                mouseX - block.x,
                mouseY - block.y
            );
            console.log('drag block', draggableBlock);
        } else {
            console.log('block not found at (' + mouseX + ',' + mouseY + ')', blocks);
        }
    });

    canvas.addEventListener('mouseup', () => {
        if (draggableBlock) {
            console.log('release block', draggableBlock);
            draggableBlock = null;
        }
    });

    canvas.addEventListener('mousemove', (e) => {
        if (draggableBlock) {
            const mouseX = e.clientX - canvas.getBoundingClientRect().left;
            const mouseY = e.clientY - canvas.getBoundingClientRect().top;

            draggableBlock.moveTo(mouseX, mouseY);

            redraw();
        }
    });
}

function findBlockBorder(blocks, x, y) {
    for (const block of blocks) {
        if (block.top == y) {
            if (block.left == x) {
                return { handle: 'topleft', block: block };
            } else if (block.right == x) {
                return { handle: 'topright', block: block };
            } else {
                return { handle: 'top', block: block };
            }
        } else if (block.right == x) {
            if (block.top == y) {
                return { handle: 'topright', block: block };
            } else if (block.bottom == y) {
                return { handle: 'bottomright', block: block };
            } else {
                return { handle: 'right', block: block };
            }
        } else if (block.bottom == y) {
            if (block.left == x) {
                return { handle: 'bottomleft', block: block };
            } else if (block.right == x) {
                return { handle: 'bottomright', block: block };
            } else {
                return { handle: 'bottom', block: block };
            }
        } else if (block.left == x) {
            if (block.top == y) {
                return { handle: 'topleft', block: block };
            } else if (block.bottom == y) {
                return { handle: 'bottomleft', block: block };
            } else {
                return { handle: 'left', block: block };
            }
        }
    }
    return null;
}

class ResizingRectangle {
    constructor(rectangle, handle) {
        this.rectangle = rectangle;
        this.handle = handle;
    }

    resizeTo(x, y) {
        this.rectangle.resizeTo(this.handle, x, y);
    }
}

function addResizeFeature(canvas, blocks, redraw) {
    var resizingBlock = null;

    canvas.addEventListener('mousedown', (e) => {
        const mouseX = e.clientX - canvas.getBoundingClientRect().left;
        const mouseY = e.clientY - canvas.getBoundingClientRect().top;

        const foundBlockBorder = findBlockBorder(blocks, mouseX, mouseY);

        if (foundBlockBorder) {
            resizingBlock = new ResizingRectangle(
                foundBlockBorder.block,
                foundBlockBorder.handle,
            );
            console.log('resize block', foundBlockBorder);
        } else {
            console.log('block border not found at (' + mouseX + ',' + mouseY + ')', blocks);
        }
    });

    canvas.addEventListener('mouseup', () => {
        if (resizingBlock) {
            console.log('release block', resizingBlock);
            resizingBlock = null;
        }
    });

    canvas.addEventListener('mousemove', (e) => {
        if (resizingBlock) {
            const mouseX = e.clientX - canvas.getBoundingClientRect().left;
            const mouseY = e.clientY - canvas.getBoundingClientRect().top;

            resizingBlock.resizeTo(mouseX, mouseY);

            redraw();
        }
    });
}

function draw(ctx, blocks) {
    ctx.clearRect(0, 0, canvas.width, canvas.height);
    for (const block of blocks) {
        drawRectangle(ctx, block);
    }
}

function init() {
    const canvas = document.getElementById("canvas");
    canvas.width = window.innerWidth;
    canvas.height = window.innerHeight;

    const ctx = canvas.getContext("2d");
    const blocks = [];

    function redraw() {
        draw(ctx, blocks)
    }

    addContextMenu(
        canvas,
        {
            onCreateNewBlock: (x, y) => {
                blocks.push(new Rectangle(x, y, 50, 50));
                console.log(blocks);
                redraw();
            }
        }
    );

    addDragFeature(canvas, blocks, () => redraw());

    addResizeFeature(canvas, blocks, () => redraw());

    redraw();
}

window.onload = init;

File: frontend/src/jsMain/kotlin/code.kt
import kotlinx.browser.document
import kotlinx.browser.window
import org.w3c.dom.CanvasRenderingContext2D
import org.w3c.dom.Element
import org.w3c.dom.HTMLButtonElement
import org.w3c.dom.HTMLCanvasElement
import org.w3c.dom.HTMLDivElement
import org.w3c.dom.Node
import org.w3c.dom.events.MouseEvent
import kotlin.math.PI
import kotlin.math.roundToInt

class Rectangle(var x: Int, var y: Int, width: Int, height: Int) {

  companion object {

    const val MIN_SIZE = 5
    const val RESIZE_HANDLE_HALF_WIDTH = 2
  }

  var width: Int
  var height: Int
  var left = x
  var right: Int
  var top = y
  var bottom: Int

  init {
    this.width = width.coerceAtLeast(MIN_SIZE)
    this.height = height.coerceAtLeast(MIN_SIZE)
    right = x + width
    bottom = y + height
  }

  fun contains(x: Int, y: Int): Boolean {
    return left < x && x < right && top < y && y < bottom
  }

  fun moveTo(x: Int, y: Int) {
    this.x = x
    this.y = y

    this.top = y
    this.bottom = y + this.height
    this.left = x
    this.right = x + this.width
  }

  enum class Handle {
    Left, TopLeft, Top, TopRight, Right, BottomRight, Bottom, BottomLeft
  }

  fun resizeTo(handle: Handle, x: Int, y: Int) {
    when (handle) {
      Handle.Left -> {
        this.width = (this.right - x).coerceAtLeast(MIN_SIZE)
        this.left = this.right - this.width
        this.x = this.left
      }

      Handle.TopLeft -> {
        resizeTo(Handle.Top, x, y)
        resizeTo(Handle.Left, x, y)
      }

      Handle.Top -> {
        this.height = (this.bottom - y).coerceAtLeast(MIN_SIZE)
        this.top = this.bottom - this.height
        this.y = this.top
      }

      Handle.TopRight -> {
        resizeTo(Handle.Top, x, y)
        resizeTo(Handle.Right, x, y)
      }

      Handle.Right -> {
        this.width = (x - this.left).coerceAtLeast(MIN_SIZE)
        this.right = this.left + this.width
      }

      Handle.BottomRight -> {
        resizeTo(Handle.Bottom, x, y)
        resizeTo(Handle.Right, x, y)
      }

      Handle.Bottom -> {
        this.height = (y - this.top).coerceAtLeast(MIN_SIZE)
        this.bottom = this.top + this.height
      }

      Handle.BottomLeft -> {
        resizeTo(Handle.Bottom, x, y)
        resizeTo(Handle.Left, x, y)
      }
    }
  }
}

class Block(
  val rectangle: Rectangle,
) {

  val connectors: MutableList<ConnectionPoint> = mutableListOf()

  fun moveTo(x: Int, y: Int) = rectangle.moveTo(x, y)
  fun resizeTo(handle: Rectangle.Handle, x: Int, y: Int) = rectangle.resizeTo(handle, x, y)
}

value class Percent private constructor(private val percent: Double) {
  constructor(part: Number, whole: Number) : this(part.toDouble() / whole.toDouble())

  infix fun of(value: Int): Int = (value * percent).roundToInt()
}

class CoordinateOnSide(
  val side: Side,
  val fromStart: Percent,
) {

  enum class Side {
    Left, Top, Right, Bottom
  }
}

class ConnectionPoint(
  val position: CoordinateOnSide,
  val block: Block,
) {

  val x
    get() = when (position.side) {
      CoordinateOnSide.Side.Left -> block.rectangle.left
      CoordinateOnSide.Side.Right -> block.rectangle.right
      CoordinateOnSide.Side.Top, CoordinateOnSide.Side.Bottom -> block.rectangle.left + (position.fromStart of block.rectangle.width)
    }

  val y
    get() = when (position.side) {
      CoordinateOnSide.Side.Top -> block.rectangle.top
      CoordinateOnSide.Side.Bottom -> block.rectangle.bottom
      CoordinateOnSide.Side.Left, CoordinateOnSide.Side.Right -> block.rectangle.top + (position.fromStart of block.rectangle.height)
    }
}

fun CanvasRenderingContext2D.draw(rectangle: Rectangle) {
  strokeRect(
      rectangle.x.toDouble(),
      rectangle.y.toDouble(),
      rectangle.width.toDouble(),
      rectangle.height.toDouble(),
  )
}

const val CONNECTION_POINT_RADIUS = 5

fun CanvasRenderingContext2D.draw(connectionPoint: ConnectionPoint, selected: ConnectionPoint?) {
  beginPath()
  arc(connectionPoint.x.toDouble(), connectionPoint.y.toDouble(),
      CONNECTION_POINT_RADIUS.toDouble(), 0.0, 2.0 * PI)
  if (connectionPoint === selected)
    fill()
  else
    stroke()
}

fun CanvasRenderingContext2D.draw(block: Block, selectedConnectionPoint: ConnectionPoint?) {
  draw(block.rectangle)
  block.connectors.forEach { draw(it, selectedConnectionPoint) }
}

fun CanvasRenderingContext2D.draw(blocks: List<Block>, selectedConnectionPoint: ConnectionPoint?) {
  clearRect(0.0, 0.0, canvas.width.toDouble(), canvas.height.toDouble())
  for (block in blocks) {
    draw(block, selectedConnectionPoint)
  }
}

fun CanvasRenderingContext2D.draw(blocksAndLinks: BlocksAndLinks) {
  save()
  translate(0.5, 0.5)
  draw(blocksAndLinks.blocks, blocksAndLinks.newLinkContext?.start)
  draw(blocksAndLinks.links)
  blocksAndLinks.newLinkContext?.let { draw(it) }
  draw(blocksAndLinks.contextMenu)
  restore()
}

fun CanvasRenderingContext2D.draw(newLinkContext: NewLinkContext) {
  beginPath()
  moveTo(newLinkContext.start.x.toDouble(), newLinkContext.start.y.toDouble())
  lineTo(newLinkContext.endX.toDouble(), newLinkContext.endY.toDouble())
  stroke()
}

fun CanvasRenderingContext2D.draw(connections: List<Link>) {
  connections.forEach { (start, end) ->
    beginPath()
    moveTo(start.x.toDouble(), start.y.toDouble())
    lineTo(end.x.toDouble(), end.y.toDouble())
    stroke()
  }
}

typealias Link = Pair<ConnectionPoint, ConnectionPoint>

class BlocksAndLinks(
  val blocks: MutableList<Block> = mutableListOf(),
  val links: MutableList<Link> = mutableListOf(),
  var contextMenu: ContextMenu? = null,
  val contextMenuItems: MutableList<ContextMenu.Item> = mutableListOf(),
  var newLinkContext: NewLinkContext? = null,
)

class NewLinkContext(
  val start: ConnectionPoint,
  var endX: Int,
  var endY: Int,
)

fun init() {
  val canvasElement = checkNotNull(document.getElementById("canvas")) {
    "element with id 'canvas' is expected"
  }
  val canvas = checkNotNull(canvasElement as? HTMLCanvasElement) {
    "element with id 'canvas' should be of type HtmlCanvasElement"
  }
  canvas.width = window.innerWidth
  canvas.height = window.innerHeight

  val renderingContext = checkNotNull(canvas.getContext("2d")) {
    "failed to obtain context '2d' from the HtmlCanvasElement with id 'canvas'"
  }
  val ctx = checkNotNull(renderingContext as? CanvasRenderingContext2D) {
    "failed to convert RenderingContext to CanvasRenderingContext2D"
  }
  val blocksAndLinks = buildBlocksAndLinks()
  val blocks = blocksAndLinks.blocks

  fun redraw() = ctx.draw(blocksAndLinks)

  addContextMenuFeature(canvas, blocksAndLinks, ::redraw)

  addCreateNewBlockFeature(blocksAndLinks, ::redraw)

  setCursorToDefaultOnMove(canvas)

  addDragFeature(canvas, blocks, ::redraw)

  addResizeFeature(canvas, blocks, ::redraw)

  addLinkFeature(canvas, blocksAndLinks, ::redraw)

  redraw()
}

fun addLinkFeature(canvas: HTMLCanvasElement, blocksAndLinks: BlocksAndLinks, redraw: () -> Unit) {
  // right-click -> check if border -> add context menu action "create connection"
  canvas.addEventListener("contextmenu", { e ->
    e as MouseEvent

    val x = e.x(canvas)
    val y = e.y(canvas)

    val border = findBlockBorder(blocksAndLinks.blocks, x, y)
    val position = border?.toCoordinateOnSide(x, y)

    if (position != null) {
      // "create connection" -> create connection point -> start "creating connection" procedure
      blocksAndLinks.contextMenuItems.add(object : ContextMenu.Item {
        override fun createElement(contextMenuDiv: HTMLDivElement): Node {
          val button = document.createElement("button") as HTMLButtonElement
          button.textContent = "Add connection point"
          button.addEventListener("click", { e ->
            e as MouseEvent
            val block = border.block
            val connectionPoint = ConnectionPoint(position, block)
            block.connectors.add(connectionPoint)
            blocksAndLinks.newLinkContext =
                NewLinkContext(connectionPoint, e.x(canvas), e.y(canvas))
            blocksAndLinks.contextMenuItems.remove(this)
          })
          return button
        }
      })
      redraw()
    }
  })

  // "create connection" procedure:
  // - draw a line between "selected connection point" and "mouse"
  canvas.addEventListener("mousemove", { e ->
    blocksAndLinks.newLinkContext?.apply {
      e as MouseEvent

      endX = e.x(canvas)
      endY = e.y(canvas)

      redraw()
    }
  })
  // - "click" -> check if border -> add "connection point" -> create link between "selected connection point" and "connection point"
  canvas.addEventListener("click", { e ->
    blocksAndLinks.newLinkContext?.also { newLinkContext ->
      e as MouseEvent

      val x = e.x(canvas)
      val y = e.y(canvas)

      val border = findBlockBorder(blocksAndLinks.blocks, x, y)
      if (border?.block != null && border.block !== newLinkContext.start.block) {
        val coordinateOnSide = border.toCoordinateOnSide(x, y)
        val connectionPoint = ConnectionPoint(coordinateOnSide, border.block)
        border.block.connectors.add(connectionPoint)
        blocksAndLinks.links.add(Link(newLinkContext.start, connectionPoint))
        blocksAndLinks.newLinkContext = null
        redraw()
      }
    }
  })
}

private fun ResizingRectangle.toCoordinateOnSide(x: Int, y: Int): CoordinateOnSide {
  val rectangle = block.rectangle
  return when (handle) {
    Rectangle.Handle.Left, Rectangle.Handle.TopLeft ->
      CoordinateOnSide(CoordinateOnSide.Side.Left, Percent(y - rectangle.y, rectangle.height))

    Rectangle.Handle.Top, Rectangle.Handle.TopRight ->
      CoordinateOnSide(CoordinateOnSide.Side.Top, Percent(x - rectangle.x, rectangle.width))

    Rectangle.Handle.Right, Rectangle.Handle.BottomRight ->
      CoordinateOnSide(CoordinateOnSide.Side.Right, Percent(y - rectangle.y, rectangle.height))

    Rectangle.Handle.Bottom, Rectangle.Handle.BottomLeft ->
      CoordinateOnSide(CoordinateOnSide.Side.Bottom, Percent(x - rectangle.x, rectangle.width))
  }
}

fun MouseEvent.x(element: Element): Int =
    (clientX - element.getBoundingClientRect().left).roundToInt()

fun MouseEvent.y(element: Element): Int =
    (clientY - element.getBoundingClientRect().top).roundToInt()

private fun buildBlocksAndLinks(): BlocksAndLinks {
  //region test data
  val block1 = Block(Rectangle(300, 300, 100, 100))
  block1.connectors.add(
      ConnectionPoint(CoordinateOnSide(CoordinateOnSide.Side.Top, Percent(3, 10)), block1))
  block1.connectors.add(
      ConnectionPoint(CoordinateOnSide(CoordinateOnSide.Side.Bottom, Percent(6, 10)), block1))
  block1.connectors.add(
      ConnectionPoint(CoordinateOnSide(CoordinateOnSide.Side.Left, Percent(12.5, 100)), block1))
  val connectionPoint1 = ConnectionPoint(
      CoordinateOnSide(CoordinateOnSide.Side.Right, Percent(87, 100)), block1)
  block1.connectors.add(connectionPoint1)

  val block2 = Block(Rectangle(600, 600, 100, 100))
  block1.connectors.add(
      ConnectionPoint(CoordinateOnSide(CoordinateOnSide.Side.Top, Percent(3, 10)), block2))
  block1.connectors.add(
      ConnectionPoint(CoordinateOnSide(CoordinateOnSide.Side.Bottom, Percent(6, 10)), block2))
  val connectionPoint2 = ConnectionPoint(
      CoordinateOnSide(CoordinateOnSide.Side.Left, Percent(12.5, 100)), block2)
  block1.connectors.add(connectionPoint2)
  block1.connectors.add(
      ConnectionPoint(CoordinateOnSide(CoordinateOnSide.Side.Right, Percent(87, 100)), block2))
  //endregion
  return BlocksAndLinks(
      blocks = mutableListOf(block1, block2),
      links = mutableListOf(Link(connectionPoint1, connectionPoint2)),
  )
}

private fun setCursorToDefaultOnMove(canvas: HTMLCanvasElement) {
  canvas.addEventListener("mousemove", {
    canvas.style.cursor = "default"
  })
}

fun main() {
  window.onload = { init() }
}
